userspace/libsinsp/filterchecks.cpp

```diff
@@ -109,6 +109,265 @@ static inline bool str_match_start(const std::string& val, size_t len, const cha
+	m_fd_success_checks.emplace(PPME_SYSCALL_LINKAT_2_X, evt_arg_res_eq_0);
+
+	m_fd_source_checks.emplace(PPME_SYSCALL_SYMLINK_X, evt_arg_linkpath);
+	m_fd_target_checks.emplace(PPME_SYSCALL_SYMLINK_X, evt_arg_target);
+	m_fd_success_checks.emplace(PPME_SYSCALL_SYMLINK_X, evt_arg_res_eq_0);
+
+	m_fd_source_checks.emplace(PPME_SYSCALL_SYMLINKAT_X, evt_arg_linkpath);
+	m_fd_target_checks.emplace(PPME_SYSCALL_SYMLINKAT_X, evt_arg_target);
+	m_fd_success_checks.emplace(PPME_SYSCALL_SYMLINKAT_X, evt_arg_res_eq_0);
+}
+
+sinsp_filter_check* sinsp_filter_check_file::allocate_new()
+{
+	return (sinsp_filter_check*)new sinsp_filter_check_file();
+}
+
+uint8_t* sinsp_filter_check_file::extract(sinsp_evt* evt, OUT uint32_t* len, bool sanitize_strings)
+{
+	*len = 0;
+	ASSERT(evt);
+
+	m_tstr = "";
+
+	if(m_fd_path_checks.empty())
+	{
+		create_fd_checks();
+	}
+
+	// First check the success conditions.
+	auto it = m_fd_success_checks.find(evt->get_type());
+	if(it == m_fd_success_checks.end())
+	{
+		return NULL;
+	}
+
+	if(!it->second->compare(evt))
+	{
+		return NULL;
+	}
+
+	std::optional<std::string> enter_param;
+
+	switch(m_field_id)
+	{
+	case TYPE_PATH:
+
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_MKDIR_X:
+		case PPME_SYSCALL_RMDIR_X:
+		case PPME_SYSCALL_UNLINK_X:
+			enter_param = evt->get_enter_evt_param("path");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		case PPME_SYSCALL_UNLINKAT_X:
+			enter_param = evt->get_enter_evt_param("name");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_file_path(evt, m_fd_path_checks);
+		};
+
+		break;
+	case TYPE_SOURCE:
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_LINK_X:
+		case PPME_SYSCALL_LINKAT_X:
+			enter_param = evt->get_enter_evt_param("oldpath");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_file_path(evt, m_fd_source_checks);
+		};
+		break;
+	case TYPE_TARGET:
+
+		// For some event types we need to get the values from the enter event instead.
+		switch(evt->get_type())
+		{
+		case PPME_SYSCALL_LINK_X:
+		case PPME_SYSCALL_LINKAT_X:
+			enter_param = evt->get_enter_evt_param("newpath");
+			if(!enter_param.has_value())
+			{
+				return NULL;
+			}
+			m_tstr = enter_param.value();
+			break;
+		default:
+			m_tstr = extract_file_path(evt, m_fd_target_checks);
+		};
+
+		break;
+	}
+
+	if(m_tstr.empty() || m_tstr == "<NA>")
+	{
+		return NULL;
+	}
+
+	// If the path is not absolute, prepend the cwd of the
+	// threadinfo to the path.
+	if(m_tstr.front() != '/')
+	{
+		sinsp_threadinfo* tinfo = evt->get_thread_info();
+
+		m_tstr = tinfo->get_cwd() + m_tstr;
```

Comment:
```suggestion
                 std::filesystem::path tstr = tinfo->get_cwd() + m_tstr;
		m_tstr = std::filesystem::absolute(tstr).lexically_normal();
```
@mstemm I think we should call lexically normal on the composite path to avoid situations like below from evading rules:
```
root@lrishi-sdm:~# pwd
/root
root@lrishi-sdm:~# ln -s /home/lrishi/td/tf ../home/lrishi/lf
# /root/../home/lrishi/lf
```

Reply:
yeah I can do that. Yay for c++17 and std::filesystem!